### PR TITLE
*: add trend api.

### PR DESCRIPTION
**Pull request by disksing**
It's for visualization of cluster's activities.

Usage example:
```
$ curl -s "http://127.0.0.1:2379/pd/api/v1/trend?from=1513086582"  // unix epoch.
{
  "stores": [
    {
      "id": 1,
      "address": "127.0.0.1:20161",
      "state_name": "Up",  // Up/Down/Offline/Disconnect
      "capacity": 5368709120,  // in bytes.
      "available": 5279425783, // in bytes.
      "region_count": 64,
      "leader_count": 20,
      "start_ts": "2017-12-19T15:18:20+08:00",
      "last_heartbeat_ts": "2017-12-19T15:20:00.874237675+08:00",
      "uptime": "1m40.874237675s",
      "hot_write_flow": 89884,   // sum of all hot regions's write flows, in bytes.
      "hot_write_region_flows": [  // a list of region's write flows, in bytes.
        89884
      ],
      "hot_read_flow": 0,
      "hot_read_region_flows": null
    },
    {
      ...
    },
   
  ],
  "history": {
    "start": 1513086582,  // start time of the history.
    "end": 1513086616,    // end time of the history.
    "entries": [
      {
        "from": 4,    // source store id.
        "to": 1033,  // target store id.
        "kind": "region",  // resource kind, "region" or "leader".
        "count": 9   // total count of resource during the time.
      },
      {
        "from": 5,
        "to": 1033,
        "kind": "leader",
        "count": 8
      },
      {
        ...
      }
    ]
  }
}
```

### CODE REVIEW

**Comment by disksing** [comment]
PTAL @Connor1996 @siddontang @nolouch

File: server/schedule/operator.go [hunk 1]
```diff
@@ -221,6 +221,46 @@ func (o *Operator) Influence(opInfluence OpInfluence, region *core.RegionInfo) {
...
+		case RemovePeer:
+			removePeerStores = append(removePeerStores, s.FromStore)
+		}
+	}
+	for i := range addPeerStores {
+		if i < len(removePeerStores) {
+			histories = append(histories, OperatorHistory{
+				FinishTime: now,
+				From:       removePeerStores[i],
+				To:         addPeerStores[i],
```
Comment (c4pt0r) [inline server/schedule/operator.go]: I think there is no order guarantee.

File: server/api/trend.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		trendStores = append(trendStores, s)
+	}
+	return trendStores, nil
+}
+
+func (h *trendHandler) getStoreFlow(stats core.StoreHotRegionsStat, storeID uint64) (storeFlow uint64, regionFlows []uint64) {
+	if stats == nil {
+		return
+	}
+	if stat, ok := stats[storeID]; ok {
```
Comment (tiancaiamao) [inline server/api/trend.go]: Not thread-safe here by design?

File: server/coordinator.go [hunk 3]
```diff
@@ -414,31 +431,18 @@ func (c *coordinator) getOperators() []*schedule.Operator {
...
+func (c *coordinator) getHistory(start time.Time) []schedule.OperatorHistory {
 	c.RLock()
 	defer c.RUnlock()
-
-	var operators []*schedule.Operator
-	for _, elem := range c.histories.Elems() {
-		op := elem.Value.(*schedule.Operator)
-		if op.Kind()&mask != 0 {
-			operators = append(operators, op)
+	var histories []schedule.OperatorHistory
```
Comment (tiancaiamao) [inline server/coordinator.go]:
```
histories := make( []schedule.OperatorHistory, 0, c.histories.Len())
```

Len() is O(1)

File: server/api/trend.go (hunk 2 above)
Comment (disksing) [reply server/api/trend.go]: It's safe as each request are accessing different copy.

File: server/schedule/operator.go (hunk 1 above)
Comment (disksing) [reply server/schedule/operator.go]: Yes, in fact we do not care about the order, only the changing trend.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**